Panic on job plan with Nomad 1.6.0

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Nomad issue
tracker reserved for bug reports and feature requests. For general usage
questions, please see: https://www.nomadproject.io/community

-->

### Nomad version
1.6.0

### Operating system and Environment details
Debian Buster

### Issue
When running a nomad plan, we get the following error message from the CLI:
```
$ ~/bin/nomad-1.6.0  job plan -hcl1 /tmp/20230719144447/nomad.hcl.final
Error during plan: Put "http://allyrion:4646/v1/job/savagecloud-monitoring-integration/plan?namespace=savagecloud-monitoring&region=integration": EOF

```
When we inspect the server logs, we see this:
```
2023-07-19T15:01:28.391-0700 [TRACE] nomad.job: job mutate results: mutator=canonicalize warnings=[] error=<nil>
2023-07-19T15:01:28.391-0700 [TRACE] nomad.job: job mutate results: mutator=connect warnings=[] error=<nil>
2023-07-19T15:01:28.391-0700 [TRACE] nomad.job: job mutate results: mutator=expose-check warnings=[] error=<nil>
2023-07-19T15:01:28.391-0700 [TRACE] nomad.job: job mutate results: mutator=constraints warnings=[] error=<nil>
2023-07-19T15:01:28.391-0700 [DEBUG] http: request complete: method=PUT path="/v1/job/savagecloud-monitoring-integration/plan?namespace=savagecloud-monitoring®ion=integration" duration=6.007368ms
2023-07-19T15:01:28.392-0700 [ERROR] http: http: panic serving 172.17.225.60:33158: runtime error: invalid memory address or nil pointer dereference
goroutine 916881 [running]:
net/http.(*conn).serve.func1()
	net/http/server.go:1854 +0xbf
panic({0x2bdf8a0, 0x50386e0})
	runtime/panic.go:890 +0x263
github.com/hashicorp/nomad/nomad.jobNodePoolMutatingHook.Mutate({0x0?}, 0xc00384e000)
	github.com/hashicorp/nomad/nomad/job_endpoint_hook_node_pool_ent.go:86 +0x5d
github.com/hashicorp/nomad/nomad.(*Job).admissionMutators(0xc000a9eaa0, 0x29e4960?)
	github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:99 +0x122
github.com/hashicorp/nomad/nomad.(*Job).admissionControllers(0xc000aba2c0?, 0x31ced0f?)
	github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:81 +0x27
github.com/hashicorp/nomad/nomad.(*Job).Plan(0xc000a9eaa0, 0xc00792f740, 0xc002b5ab60)
	github.com/hashicorp/nomad/nomad/job_endpoint.go:1747 +0x196
reflect.Value.call({0xc000194e40?, 0xc0009193f8?, 0x7fef9c699c40?}, {0x31cfab3, 0x4}, {0xc003663378, 0x3, 0x0?})
	reflect/value.go:586 +0xb0b
reflect.Value.Call({0xc000194e40?, 0xc0009193f8?, 0x30925a0?}, {0xc003663378?, 0xc0036633c8?, 0xa594b8?})
	reflect/value.go:370 +0xbc
net/rpc.(*service).call(0xc0003f39c0, 0x2f32440?, 0xc003663450?, 0x0, 0xc0005b3280, 0x40?, {0x30925a0?, 0xc00792f740?, 0x0?}, {0x2d55360, ...}, ...)
	net/rpc/server.go:382 +0x226
net/rpc.(*Server).ServeRequest(0x70?, {0x38eb430, 0xc0032ef640})
	net/rpc/server.go:503 +0x18c
github.com/hashicorp/nomad/nomad.(*Server).RPC(0xc000aba2c0, {0x31e3f05, 0x8}, {0x30925a0?, 0xc00792f680}, {0x2d55360?, 0xc002b5aaf0})
	github.com/hashicorp/nomad/nomad/server.go:1925 +0xf3
github.com/hashicorp/nomad/command/agent.(*Agent).RPC(0xc000ee72d0?, {0x31e3f05?, 0xc0005aa500?}, {0x30925a0?, 0xc00792f680?}, {0x2d55360?, 0xc002b5aaf0?})
	github.com/hashicorp/nomad/command/agent/agent.go:1216 +0x45
github.com/hashicorp/nomad/command/agent.(*HTTPServer).jobPlan(0xc000ee72d0, {0x38e98d0, 0xc00792f2c0}, 0xc0005aa500, {0xc005075f1c, 0x22})
	github.com/hashicorp/nomad/command/agent/job_endpoint.go:181 +0x31e
github.com/hashicorp/nomad/command/agent.(*HTTPServer).JobSpecificRequest(0x412ac7?, {0x38e98d0, 0xc00792f2c0}, 0xc0005aa500)
	github.com/hashicorp/nomad/command/agent/job_endpoint.go:82 +0x41b
github.com/hashicorp/nomad/command/agent.(*HTTPServer).auditHandler.func1({0x38e98d0?, 0xc00792f2c0}, 0xc0005aa400)
	github.com/hashicorp/nomad/command/agent/http_ent.go:239 +0x1fa
github.com/hashicorp/nomad/command/agent.(*HTTPServer).wrap.func1({0x38e98d0, 0xc00792f2c0}, 0xc0005aa400)
	github.com/hashicorp/nomad/command/agent/http.go:710 +0x17b
net/http.HandlerFunc.ServeHTTP(0xc001b93dd0?, {0x38e98d0?, 0xc00792f2c0?}, 0x510e87?)
	net/http/server.go:2122 +0x2f
net/http.(*ServeMux).ServeHTTP(0x0?, {0x38e98d0, 0xc00792f2c0}, 0xc0005aa400)
	net/http/server.go:2500 +0x149
github.com/gorilla/handlers.CompressHandlerLevel.func1({0x38e7a10?, 0xc0017602a0}, 0xc0005aa400)
	github.com/gorilla/handlers@v1.5.1/compress.go:141 +0x6d0
net/http.HandlerFunc.ServeHTTP(0xc005075f44?, {0x38e7a10?, 0xc0017602a0?}, 0x74cc14?)
	net/http/server.go:2122 +0x2f
net/http.serverHandler.ServeHTTP({0x38db890?}, {0x38e7a10, 0xc0017602a0}, 0xc0005aa400)
	net/http/server.go:2936 +0x316
net/http.(*conn).serve(0xc0036f3dd0, {0x38ea358, 0xc00140d260})
	net/http/server.go:1995 +0x612
created by net/http.(*Server).Serve
	net/http/server.go:3089 +0x5ed
```

The 1.6.0 leader logs also show lots of these:
```
2023-07-20T11:59:15.945-0700 [ERROR] nomad: failed to fetch node pools from authoritative region: error="rpc error: rpc: can't find service NodePool.List"
2023-07-20T11:59:15.946-0700 [ERROR] nomad: failed to fetch ACL auth-methods from authoritative region: error="rpc error: rpc: can't find method ACL.ListAuthMethods"
2023-07-20T11:59:15.953-0700 [ERROR] nomad: failed to fetch ACL binding rules from authoritative region: error="rpc error: rpc: can't find method ACL.ListBindingRules"
2023-07-20T11:59:46.014-0700 [ERROR] nomad: failed to fetch node pools from authoritative region: error="rpc error: rpc: can't find service NodePool.List"
2023-07-20T11:59:46.015-0700 [ERROR] nomad: failed to fetch ACL auth-methods from authoritative region: error="rpc error: rpc: can't find method ACL.ListAuthMethods"
2023-07-20T11:59:46.022-0700 [ERROR] nomad: failed to fetch ACL binding rules from authoritative region: error="rpc error: rpc: can't find method ACL.ListBindingRules"
2023-07-20T12:00:16.083-0700 [ERROR] nomad: failed to fetch ACL auth-methods from authoritative region: error="rpc error: rpc: can't find method ACL.ListAuthMethods"
2023-07-20T12:00:16.084-0700 [ERROR] nomad: failed to fetch node pools from authoritative region: error="rpc error: rpc: can't find service NodePool.List"
2023-07-20T12:00:16.090-0700 [ERROR] nomad: failed to fetch ACL binding rules from authoritative region: error="rpc error: rpc: can't find method ACL.ListBindingRules"
```

--- a/.changelog/18017.txt
+++ b/.changelog/18017.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+core: Fixed a bug where namespace were not canonicalize on snapshot restore, resulting in potential nil access panic
+```

--- a/.changelog/18017.txt
+++ b/.changelog/18017.txt
@@ -1,3 +1,3 @@
 ```release-note:bug
-core: Fixed a bug where namespace were not canonicalize on snapshot restore, resulting in potential nil access panic
+core: Fixed a bug where namespaces were not canonicalized on snapshot restore, resulting in potential nil access panic
 ```

--- a/nomad/state/state_store_restore.go
+++ b/nomad/state/state_store_restore.go
@@ -205,6 +205,8 @@ func (r *StateRestore) ScalingEventsRestore(jobEvents *structs.JobScalingEvents)
 
 // NamespaceRestore is used to restore a namespace
 func (r *StateRestore) NamespaceRestore(ns *structs.Namespace) error {
+	// Handle upgrade path.
+	ns.Canonicalize()
 	if err := r.txn.Insert(TableNamespaces, ns); err != nil {
 		return fmt.Errorf("namespace insert failed: %v", err)
 	}

